### DISCUSSION TITLE
Write mfront behaviours using external source code efficiently

### DISCUSSION THREAD

**Discussion by pbouda95** [thelfer/tfel#246] [Q&A]
Dear all,

Does somebody already tried to implement a behaviour using c++ external source files ?
In my case I would like to use a personal struct that handles complex calculations ...

The documentation seems not to be very talkative about this option of MFront so I do not know how to start ...

Thanks in advance for the answers/tips,

Cheers,
Pascal

**Comment by thelfer** [comment]
@pbouda95 You can easily use external C++ files using the following keywords:

- `@Includes`: see `mfront --help-keyword=Implicit:@Includes`
- `@Sources`: see `mfront --help-keyword=Implicit:@Sources`
- `@Link`: see `mfront --help-keyword=Implicit:@Link`

Note that you may want to specify the `@Link` keyword on the command line as follows:

```bash
$ mfront --obuild --interface=generic --@Link='"-lm"' SaintVenantKirchhoffElasticity.mfront
```

**Reply by pbouda95** [reply]
Hi @thelfer ,

Thanks for that ! 

To check if this can meet my needs, I did try to include a small header at the beginning of my mfront file owing to the `@Include` keyword as follows:
`@Includes{
#include"foo.hxx"
};`

The header file is as follows:
`class Foo {
public:
 void foo();
 void bar();
};`

But when compiling, it throws the following error:
`In file included from template.mfront:11,
                 from ../include/TFEL/Material/DPDC.hxx:33,
                 from DPDC-generic.cxx:43:
/home/pb245956/volatile/Europlexus/2022/dpdc/Foo.hxx:1:7: error: redefinition of ‘class Foo’
    1 | class Foo {
      |       ^~~
compilation terminated due to -Wfatal-errors.
`

Did I miss something ? I tried to look at the generated files but things seems a bit foggy for me ...

Cheers,

**Reply by thelfer** [reply]
Your file `foo.hxx` is ill-formed. You need to add header guards (see <https://en.wikipedia.org/wiki/Include_guard>) as follows:

~~~~{.cxx}
#ifndef LIB_FOO_HHX
#define LIB_FOO_HXX

class Foo { public: void foo(); void bar(); };

#endif /* LIB_FOO_HXX */
~~~~

You can also use `#pragma once` which is non standard but supported by all compilers (see <https://en.wikipedia.org/wiki/Pragma_once>).

**Reply by pbouda95** [reply]
Argh ok ! I did not know that the header is to be called multiple times. 

It works like a charm now ... Thanks !

**Comment by pbouda95** [comment]
Hello @thelfer,

I would like now to use some of your abstract types such as ```stensor``` or ```real``` inside my external classes. What is the best way to do that ?

Cheers,

**Reply by thelfer** [reply]
Simply use template function as follows:

```cxx
template<unsigned short N, typename real>
auto do_things(const tfel::math::stensor<N,real>& sig){
....
}
```

**Reply by thelfer** [reply]
BTW, there is almost no virtual inheritance in `TFEL/Math` and `TFEL/Material`.

**Comment by fpguillet** [comment]
It is not clear to me how to include external C++ source files where, for example, some functions are implemented. I understand the @Includes statement where headers are loaded in the .mfront file. But I wonder about the implementation of the functions which I'd like to put in .cpp file(s) and not in a @Sources statement. Or is it possible to use mfront in this  way ;
mfront mybehaviour.mfront someadditionalfile.cpp
?

Thanks,
François.

**Reply by thelfer** [reply]
Hi François,

The easiest way is to do:

```cxx
@Sources{
#include "someadditionalfile.cpp"
}
```
